### ISSUE TITLE
关于训练和评估时候的图像尺寸问题

### ISSUE THREAD

**Issue by David-dotcom666** [PaddlePaddle/PaddleSeg#3867]
### 问题确认 Search before asking

- [X] 我已经搜索过问题，但是没有找到解答。I have searched the question and found no related answer.

### 请提出你的问题 Please ask your question

1.我在train_dataset的transforms使用了    
- type: Resize
      target_size: [ 512, 512 ]
那么我在val_dataset也需要
- type: Resize
      target_size: [ 512, 512 ]
这样才能保持训练和评估尺寸的一致性吗？
2.如果我在rain_dataset的transforms使用了    
    - type: ResizeStepScaling
      min_scale_factor: 0.5
      max_scale_factor: 0.5
      scale_step_size: 0
那么这个时候我在val_dataset中的transforms要怎么设置呢？我试了也设置成这样，在评估的时候会报标注和pred尺寸不一致。如果在val_dataset不设置尺寸倒是可以，但是这时候评估的是原图尺寸上的评估吧？我想在评估也进行自动一半尺寸的。这时候网络的输入尺寸是多少呢？是原图吗？导出模型尺寸设置多少合适？

我目前的需求是可以训练不同尺寸的图，且可以选择训练精度。例如0.5，我就使用ResizeStepScaling到固定的0.5，然后我在导出后的模型预处理的时候先把图像进行ResizeStepScaling，然后padding到导出的尺寸。

**Comment by Bobholamovic** [comment]
你好，训练和评估的尺寸并不一定要一模一样；通常在训练时可以进行随机resize，但评估阶段需要resize到一个固定的尺寸。如果不进行设置，通常评估时将使用原图尺寸。具体的训练、验证输入尺寸选择取决于您使用的模型和数据集，并没有一个标准答案。通常，分割模型在模型输入尺寸上具备一定的泛化性能。

**Comment by David-dotcom666** [comment]
> 你好，训练和评估的尺寸并不一定要一模一样；通常在训练时可以进行随机resize，但评估阶段需要resize到一个固定的尺寸。如果不进行设置，通常评估时将使用原图尺寸。具体的训练、验证输入尺寸选择取决于您使用的模型和数据集，并没有一个标准答案。通常，分割模型在模型输入尺寸上具备一定的泛化性能。
谢谢回答，我想请教一个问题，我看市面上有一些商业软件做到了样本数量很少，甚至只有1.2张的情况下训练出的效果也还不错。我想请教下，对于这种情况，基本上样本都是不平衡的而且差的非常多，什么样的loss组合会比较泛用呢？我看了一些帖提到CE或者multiclass的，那么这些我在使用的时候需要加权重吗？最近在测试pplite-seg，它默认使用三个loss，三个loss是平等的吗

**Comment by Bobholamovic** [comment]
在样本类别非常不平衡的情况下，一种常见且有效的做法是在损失函数中考虑这一点，例如为多类的CE loss添加类别权重。以PP-LiteSeg为例，配置文件[`pp_liteseg_stdc1_cityscapes_1024x512_scale0.5_160k.yml`](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.10/configs/pp_liteseg/pp_liteseg_stdc1_cityscapes_1024x512_scale0.5_160k.yml) 中对损失函数的配置如下：

```yaml
loss:
  types:
    - type: OhemCrossEntropyLoss
      min_kept: 130000   # batch_size * 1024 * 512 // 16
    - type: OhemCrossEntropyLoss
      min_kept: 130000
    - type: OhemCrossEntropyLoss
      min_kept: 130000
  coef: [1, 1, 1]
```

该模型在训练阶段给出3个输出（1个主要输出和2个辅助输出），以上配置对每个输出都使用`OhemCrossEntropyLoss`和标签计算损失，得到的三个loss以`coef`指定的系数`[1, 1, 1]`加权求和。要想对其中的每个损失函数增加类别权重，可以将配置修改如下：

```yaml
loss:
  types:
    - type: OhemCrossEntropyLoss
      min_kept: 130000   # batch_size * 1024 * 512 // 16
      weight: [0.5, 0.2, 0.3]
    - type: OhemCrossEntropyLoss
      min_kept: 130000
      weight: [0.5, 0.2, 0.3]
    - type: OhemCrossEntropyLoss
      min_kept: 130000
      weight: [0.5, 0.2, 0.3]
  coef: [1, 1, 1]
```

以上的设置将（对于每个损失函数）为第0类添加0.5的权重，为第1类和第2类分别添加0.2和0.3的权重。

**Comment by David-dotcom666** [comment]
> 在样本类别非常不平衡的情况下，一种常见且有效的做法是在损失函数中考虑这一点，例如为多类的CE loss添加类别权重。以PP-LiteSeg为例，配置文件[`pp_liteseg_stdc1_cityscapes_1024x512_scale0.5_160k.yml`](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.10/configs/pp_liteseg/pp_liteseg_stdc1_cityscapes_1024x512_scale0.5_160k.yml) 中对损失函数的配置如下：
> 
> ```yaml
> loss:
>   types:
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000   # batch_size * 1024 * 512 // 16
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>   coef: [1, 1, 1]
> ```
> 
> 该模型在训练阶段给出3个输出（1个主要输出和2个辅助输出），以上配置对每个输出都使用`OhemCrossEntropyLoss`和标签计算损失，得到的三个loss以`coef`指定的系数`[1, 1, 1]`加权求和。要想对其中的每个损失函数增加类别权重，可以将配置修改如下：
> 
> ```yaml
> loss:
>   types:
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000   # batch_size * 1024 * 512 // 16
>       weight: [0.5, 0.2, 0.3]
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>       weight: [0.5, 0.2, 0.3]
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>       weight: [0.5, 0.2, 0.3]
>   coef: [1, 1, 1]
> ```
> 
> 以上的设置将（对于每个损失函数）为第0类添加0.5的权重，为第1类和第2类分别添加0.2和0.3的权重。

歪下楼，顺便贴下我调参的一些经历给有缘人参考吧：

- 1. 关于几种常见的自动计算权重的方式：
- 1.1.softmax，这种方式的出处是ppseg contirb中的：
```
freq_num = np.asarray(list(overall_class_stats.values()), dtype='float64')//标注所有类型的像素个数集合
freq = freq_num / np.sum(freq_num)
freq = 1 - freq
freq = F.softmax(paddle.to_tensor(freq / temperature, place='cpu'), axis=-1)
```
- 1.2.以类别像素数量最小的方式
```
min_freq = np.min(freq_num)
class_weights_new = 1 / (freq_num / min_freq)
```
- 1.3.像素数量中位数方式
```
median_freq = np.median(freq_num)
class_weights_new = median_freq / freq_num
class_weights_new /= np.sum(class_weights_new)
```
- 1.4.sklearn 库的compute_class_weight
- 2.如果样本极度不平衡的情况下，设置权重可能更加困难，甚至不如不设置。目前我尝试的都是去降低背景像素的比例或者计算
- 3.如果使用了CE Loss，在极度不均衡的情况下，我使用了top_k_percent_pixels，计算检测目标的像素和背景像素比重后*2。如果是mixloss就再降低CE Loss的比例，可以很大程度的缓解极不平衡训练问题，尤其是使用CE+LovaszSoftmaxLoss训练快速收敛后，miou快速下降的问题。

**Comment by David-dotcom666** [comment]
> 在样本类别非常不平衡的情况下，一种常见且有效的做法是在损失函数中考虑这一点，例如为多类的CE loss添加类别权重。以PP-LiteSeg为例，配置文件[`pp_liteseg_stdc1_cityscapes_1024x512_scale0.5_160k.yml`](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.10/configs/pp_liteseg/pp_liteseg_stdc1_cityscapes_1024x512_scale0.5_160k.yml) 中对损失函数的配置如下：
> 
> ```yaml
> loss:
>   types:
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000   # batch_size * 1024 * 512 // 16
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>   coef: [1, 1, 1]
> ```
> 
> 该模型在训练阶段给出3个输出（1个主要输出和2个辅助输出），以上配置对每个输出都使用`OhemCrossEntropyLoss`和标签计算损失，得到的三个loss以`coef`指定的系数`[1, 1, 1]`加权求和。要想对其中的每个损失函数增加类别权重，可以将配置修改如下：
> 
> ```yaml
> loss:
>   types:
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000   # batch_size * 1024 * 512 // 16
>       weight: [0.5, 0.2, 0.3]
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>       weight: [0.5, 0.2, 0.3]
>     - type: OhemCrossEntropyLoss
>       min_kept: 130000
>       weight: [0.5, 0.2, 0.3]
>   coef: [1, 1, 1]
> ```
> 
> 以上的设置将（对于每个损失函数）为第0类添加0.5的权重，为第1类和第2类分别添加0.2和0.3的权重。
@Bobholamovic 
想请教下，我使用的ppliteseg，我想尝试backbone_indices从默认的[2,3,4]改成[1,2,3]，那么我的arm_out_chs和seg_head_inter_chs等需要变更吗

**Comment by Bobholamovic** [comment]
通常需要，因为选取不同层的特征，可能导致特征通道数的改变，具体可以参考backbone的实现代码

**Comment by David-dotcom666** [comment]
> 通常需要，因为选取不同层的特征，可能导致特征通道数的改变，具体可以参考backbone的实现代码
@Bobholamovic 
我使用了   backbone_indices=[1, 2, 3],
                 arm_type='UAFM_SpAtten',
                 cm_bin_sizes=[1, 2, 4],
                 cm_out_ch=128,
                 arm_out_chs=[32, 32, 64],
                 seg_head_inter_chs=[32, 32, 32],
              训练起来几乎不收敛，能帮看看应该怎么配吗？

**Comment by Bobholamovic** [comment]
你用的具体是哪个配置呀

**Comment by David-dotcom666** [comment]
> 你用的具体是哪个配置呀
用的STDC2的，configs/pp_liteseg/pp_liteseg_stdc2_cityscapes_1024x512_scale1.0_160k.yml
模型这里改成了
```
model:
  type: PPLiteSeg
  backbone:
    type: STDC2
    pretrained: https://bj.bcebos.com/paddleseg/dygraph/PP_STDCNet2.tar.gz
  backbone_indices=[1, 2, 3],
  arm_out_chs: [32, 32, 64]
  seg_head_inter_chs: [32, 32, 32]
```
改了几种都是不收敛了，是不是要从头训练了....

**Comment by Bobholamovic** [comment]
对于这一配置，将`backbone_indices`调整为`[1, 2, 3]`后，模型后续部分使用的backbone特征通道数从`[256, 512, 1024]`变成了`[64, 256, 512]`，建议将`arm_out_chs`（arm模块使用的中间通道数和输出通道数）和`seg_head_inter_chs`（分割头使用的中间通道数）也对应调小，将其分别设置为`[32, 32, 64]`和`[32, 32, 32]`是一个合理的选择。这里应该没有bug，请问你说的“不收敛”以及“从头训练”具体是指什么呢？调整模型结构后通常都需要重新训练，考虑以旧结构的模型权重参数进行部分参数初始化可能可以加快收敛速度。

**Comment by David-dotcom666** [comment]
> 对于这一配置，将`backbone_indices`调整为`[1, 2, 3]`后，模型后续部分使用的backbone特征通道数从`[256, 512, 1024]`变成了`[64, 256, 512]`，建议将`arm_out_chs`（arm模块使用的中间通道数和输出通道数）和`seg_head_inter_chs`（分割头使用的中间通道数）也对应调小，将其分别设置为`[32, 32, 64]`和`[32, 32, 32]`是一个合理的选择。这里应该没有bug，请问你说的“不收敛”以及“从头训练”具体是指什么呢？调整模型结构后通常都需要重新训练，考虑以旧结构的模型权重参数进行部分参数初始化可能可以加快收敛速度。

不收敛是说，我使用默认配置训练我的数据可以收敛，但是改成
```
 backbone_indices=[1, 2, 3],
  arm_out_chs: [32, 32, 64]
  seg_head_inter_chs: [32, 32, 32]
```
后训练了比原来多一倍的iter，miou还都是0
重新训练是说我要在imagenet或者COCO上从头开始吗？

**Comment by Bobholamovic** [comment]
正常来说不应该出现这个现象～ 请问你在训练程序启动时有注意到什么不寻常的警告信息吗？另外，想确认下` backbone_indices=[1, 2, 3],`是你实际的写法吗？这样应该是不符合YAML语法规则的

**Comment by David-dotcom666** [comment]
> 正常来说不应该出现这个现象～ 请问你在训练程序启动时有注意到什么不寻常的警告信息吗？另外，想确认下` backbone_indices=[1, 2, 3],`是你实际的写法吗？这样应该是不符合YAML语法规则的
1.没有什么特别的警告
2.`backbone_indices=[1, 2, 3]`,是你实际的写法吗->忽然发现这里好像不对，我再试下再回
->ZZ了，确实有反应了。应该是从网络复制默认参数没修改到- -!。我试试效果会不会好点
先关闭这条啦，感谢!

**Comment by David-dotcom666** [comment]
> [Jan 13, 2025](#event-15901246579)

![image](https://github.com/user-attachments/assets/4c27b0ec-7433-4122-b5d0-c6b9220d99c4)
看起来效果不怎么样

**Comment by Bobholamovic** [comment]
请问图中横轴是迭代数、纵轴是mIoU精度吗？
